src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AdvSimdHelper.cs

```diff
@@ -0,0 +1,135 @@
+
+            // Leftover control character in the ASCII range - 0x7F
+            // Since we are dealing with sbytes, 0x7F is the only value that would meet this comparison.
+            mask = AdvSimd.Or(mask, AdvSimd.CompareGreaterThan(sourceValue, s_tildeMaskSByte));
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<sbyte> CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin(Vector128<sbyte> sourceValue)
+        {
+            Debug.Assert(AdvSimd.IsSupported);
+
+            Vector128<sbyte> mask = CreateEscapingMask_UnsafeRelaxedJavaScriptEncoder(sourceValue);
+
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_ampersandMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_apostropheMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_plusSignMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_lessThanSignMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_greaterThanSignMaskSByte));
+            mask = AdvSimd.Or(mask, AdvSimd.CompareEqual(sourceValue, s_graveAccentMaskSByte));
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<short> CreateAsciiMask(Vector128<short> sourceValue)
+        {
+            Debug.Assert(AdvSimd.IsSupported);
+
+            // Anything above short.MaxValue but less than or equal char.MaxValue
+            // That's because anything between 32768 and 65535 (inclusive) will overflow and become negative.
+            Vector128<short> mask = AdvSimd.CompareLessThan(sourceValue, s_nullMaskInt16);
+
+            // Anything above the ASCII range
+            mask = AdvSimd.Or(mask, AdvSimd.CompareGreaterThan(sourceValue, s_maxAsciiCharacterMaskInt16));
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ContainsNonAsciiByte(Vector128<sbyte> value)
+        {
+            Debug.Assert(AdvSimd.Arm64.IsSupported);
+            return AdvSimd.Arm64.MinAcross(value).ToScalar() < 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetIndexOfFirstNonAsciiByte(Vector128<byte> value)
+        {
+            Debug.Assert(AdvSimd.Arm64.IsSupported && BitConverter.IsLittleEndian);
+
+            // extractedBits[i] = (value[i] & 0x80 == 0x80) & (1 << (12 * (i % 2)))
+            Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, s_mostSignficantBitMask), s_mostSignficantBitMask);
+            Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitmask);
+
+            // collapse mask to lower bits
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
```

Comment:
Couldn't we cut out the `>> 2` if this did

```
var tmp = extractedBits.AsUInt64();
var msk = AddPairwise(tmp, tmp).AsUInt32().ToScalar();
```

I believe that would give an identical mask to what x86 gives, given that you have all 1's or all 0's from the comparison and each index is set to a unique element.

Reply:
Out of curiosity, why would we want to cut out the `>> 2`? Is it expected that it could be slower than an additional add pairwise operation? Or is it just a case of avoiding `TrailingZeroCount(long)`?

> given that you have all 1's or all 0's from the comparison and each element is given a specific bit position from 0 to 15.

Note that the current approach uses `0x1001` as the bitmask, so this is not true currently. It was the case in the original approach, which however required triple `AddPairwise` application to obtain an x86 equivalent result. One possible alternative might be to do the following:

```csharp
var bitmask = Vector128.Create((uint)0x40100401).AsByte();
...
Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, bitmask);
extractedBits = AdvSimd.AddPairwise(extractedBits, extractedBits);
extractedBits = AdvSimd.AddPairwise(extractedBits, extractedBits);

int mask = extractedBits.AsInt32().ToScalar();
Debug.Assert(mask >= 0);
return BitOperations.TrailingZeroCount(mask) >> 1;
```